### PR TITLE
Implement bastion star fort, flesh out layout

### PR DESCRIPTION
**Pull request by chaosvolt**
## Summary

SUMMARY: Content "Add a bastion star fort to spawn cannons, use a more appealing layout"

## Purpose of change

So, @hirohito1 / Moth was considering porting over bastion forts from DDA so we'd have cool stuff like 3 inch ordinance guns and such, but got frustrated while testing. I took a look at it and uh:
image

I know it's hard to draw a pentagon in a grid-based map but bleh, DDA's version is basically unusable. It really suffers for being mashed into a 3x4 map footprint. While I'm reasonably certain I could grid up a less deformed shape for a 5-point fort, I opted instead to start this off by working on a layout for a more pleasantly symmetrical 4-point style.

Been basing my work of Fort Ticonderoga, minus the demi-lunes and with some adjustments to fit a square profile instead of rectangular:
image

## Describe the solution

- [X] Added mapgen file for bastion fort. Contains ground floor, second floor with ramparts, and roof level. Three buildings present, two of them partially converted into modern amenities like museum displays, restrooms, a gift shop, break room etc. The enlisted barracks meanwhile remains mostly flavored as a historical building.
- [X] Added monster spawns to the fort. At the moment it uses existing monsters, but random park goers are already a better variety and include some potential evolutionary threats compared to the bland monstergroup that DDA's version uses.
- [X] Added mapgen palette for star fort. Went with writing my own instead of using DDA's version.
- [X] Added overmap terrain entries for star fort. Based of bandit camp instead of using DDA version.
- [X] Added overmap special entry for star fort. Ported over from DDA but slightly increased occurrence chance.
- [X] Ported over 3 inch ordinance rifle from DDA. Didn't see anything wrong with the stats on it but the useless spiked guns that exist to troll the player are a flat no.
- [X] Ported over vehicle form of the cannon to place on the ramparts. I had to mostly redo it because it was...very, very wrong. Wooden frames when the gun carriage IRL weighs is minor, but they managed to screw it up so badly that vehicle UI and drag physics wise the wheels are on front and back of the cannon instead of left and right. So I had to redo the vehicle def to fix bicycle cannons.
- [X] Ported over the basic iron cannonball item and ammo from DDA. Stats are about the same aside from sanity-checking weight and volume, but description makes it clear the explosive shells are were it's at (to be added eventually).
- [X] Ported over recipes for cannonball and the resulting combined load. Quick wikipedia search says the 3-inch used a pound of gunpowder, so sanity-checked the downright insane ~~6 kilos~~ 4.5 kilos (blackpowder weighs 1.51 grams in DDA compared to 2 grams in BN) worth of powder down to 450 grams worth.
- [X] Implemented relevant itemgroups to spawn loot at the fort. Added cannon spawns to mansion and museum in the same places DDA spawns them, but when with "not 1 fucking percent" spawn rate for the mansion spawn since it's only a single potential tile in a single mansion variant. Also placed the museum cannon in a less random spot. Opted not to make mansion ammo stashes spawn cannon ammo however since that'd make them very common in all mansions when only one mansion variant can actually spawn them. Instead, added cannonballs and shells (but not assembled loads) to `museum_guns` itemgroup.

Other additions:
- [X] Added small wooden barrels, 20 liter size for sake of holding a reasonable amount of powder for cannons. Makes it look more fitting since powder kegs were meant to be hauled around by hand, and makes the smaller max charges look less weird when it's 2.5 liters out of 20 instead of 2.5/100.
- [X] Added recipe for small wooden barrel, using copper as mentioned. Idea being if you do use metal for a powder keg you want one that won't spark.
- [X] Added explosive shells for use with the cannon, set as the main ammotype for cannons. The individual shells are set to spawn occasionally in the fort.
- [X] Added recipe for crafting the explosive shells, based off other explosive recipes and generally balanced to be more effort than plain round shot.

## Describe alternatives you've considered

1. Porting the original version, then trying to fix the proportions and layout.
2. Doing basically what I'm already doing right now except pentagon shaped because I just love trying to fit a door onto a diagonal wall.
3. Making a Roman castra instead. "You know what, fuck you" *anachronizes your New England fort*
4. Making Moth do all this instead.

## Testing

1. Checked affected files for syntax and lint errors.
2. Load-tested in compiled build.
3. Debug placed a bastion fort to confirm things look right.

image
image

## Additional context

Related DDA PR (nothing original to it ported over yet), squished pentagons by @jokermatt999: https://github.com/CleverRaven/Cataclysm-DDA/pull/45215

Optional stuff we could add later on:
* Porting over reenactor zombies as well. They don't really add much aside from flavor and loot, so I think I'm content to regard them as an option addition to this.
* Makeshift canister shot using a small tin can and lead, simpler tools to make than lead shot but worse range, to be balanced as shot/HP has generally been. This would put expected damage at up to 450. Probably would not use AoE cone behavior yet because any level of damage high enough to reliably kill an NPC (keeping in mind that AoE splits the damage so they get cheese grater'd across all 6 HP pools) would be "this melts hulks by the dozen" tier when used against monsters since they only have a single HP pool.
* An M1881 Gatling gun, something Moth mused upon. Specifically that one because it uses an ammo type already present in the game, .45-70.
* Vehicle form of M1881 Gatling gun, if present set to randomly spawn in museums in place of cannons, maybe optionally injected into the bastion fort on occasion too on the basis that the fort is no longer an authentic Civil War era fort but instead a general educational and living history exhibit.
* Feral reenactors who get one chance to gank you with a musket shot before being stuck with just a bayonet charge. No idea if it's a thing in DDA or not.
* Adding a variant of gradual ramp tiles specifically for stone floors instead of using the sidewalk version.

### CODE REVIEW

File: data/json/items/ammo_types.json [hunk 1]
```diff
@@ -639,7 +639,7 @@
   {
     "type": "ammunition_type",
     "id": "cannon",
-    "name": "cannon",
+    "name": "3-inch cannon",
```
Comment (scarf005) [inline data/json/items/ammo_types.json:642]: _nooooooooo_

Comment (chaosvolt) [reply data/json/items/ammo_types.json:642]: Suffer :D

**Comment by olanti-p** [comment]
I'm getting an error when trying to examine a cannon
```
10:29:14.006 ERROR DEBUGMSG : /home/olanti/code/cata/a/src/item.cpp:628 [item &item::ammo_set(const itype_id &, int)] Tried to set invalid ammo cannonball_explosive[1] for cannon_3in_ordnance
```

**Review by scarf005 — APPROVED** [review]
Cataclysm: Bright Nights - a16e24490aff_01

https://github.com/cataclysmbnteam/Cataclysm-BN/assets/54838975/955178fa-5a5e-4602-8980-afceeea5eae9

cannons need controls to be installed manually but they're fun.

**Comment by chaosvolt** [comment]
> cannons need controls to be installed manually but they're fun.

They can be fired manually by standing on the tile that the cannon itself is installed on, and hitting `f` with your hands empty.

**Comment by scarf005** [comment]
oh, didn't knew that. could be nice to have as a guide in docs page

**Pull request merged**